### DISCUSSION TITLE
Return type of migrate in zustand persist

### DISCUSSION THREAD

**Discussion by bobdepreziz** [pmndrs/zustand#2357] [Q&A]
Hello,

I have a state for my store I persist that looks different than the store itself since it is partialized. I use migrate if something changes to that persisted state.

I am declaring the type like this:

PersistOptions<AssetHistoryStoreProps, PersistedStoreState>

In the file "persist.d.ts" is see: 

export interface PersistOptions<S, PersistedState = S>

And then the type def of merge this returns S which is the normal store state and not the persisted state:

    /**
     * A function to perform custom hydration merges when combining the stored state with the current one.
     * By default, this function does a shallow merge.
     */
    merge?: (persistedState: unknown, currentState: S) => S;
    
    Should this not be PersistedStoreState? So like this:
    
        merge?: (persistedState: unknown, currentState: S) => PersistedState;
        
Any feedback would be appreciated, thank you!

**Comment by dai-shi** [comment]
`PersistedState` is the type for the storage value, and `S` is the type for the store value in memory. So, it seems `S` is correct.

**Reply by bobdepreziz** [reply]
Actually I ment the migrate function since that one migrates the persisted state, I would expect it to return the persisted state

/**
     * A function to perform persisted state migration.
     * This function will be called when persisted state versions mismatch with the one specified here.
     */
    migrate?: (persistedState: unknown, version: number) => S | Promise<S>;

**Reply by dai-shi** [reply]
Hm, yeah, for `migrate`, it seems the return value is `PersistedState | Promise< PersistedState>`.

**Reply by bobdepreziz** [reply]
Would consider this as a bug?

**Reply by dai-shi** [reply]
Probably. Please open a PR so that someone else can review it.
Unfortunately, it wouldn't be easy to add a test, but if you have an idea, it'd be good.

**Reply by bobdepreziz** [reply]
I opened a PR see:
https://github.com/pmndrs/zustand/pull/2360

**Reply by bobdepreziz** [reply]
@dai-shi I also noticed that the type of persistedState in merge is of type unknown, is this intentional? Is there a reason why it does not get the PersistedState type?

 /**
   * A function to perform custom hydration merges when combining the stored state with the current one.
   * By default, this function does a shallow merge.
   */
  merge?: (persistedState: unknown, currentState: S) => S

**Reply by dai-shi** [reply]
That is very intentional, because we really don't know what's in the stored value. It comes from outside.

**Reply by bobdepreziz** [reply]
@dai-shi That is what I thought but was not sure, thanks for clarifying.
